### ISSUE TITLE
Help - My PC crashes when using node.js

### ISSUE THREAD

**Issue by wappevifi** [nodejs/help#1045]
* **Node.js Version**: not sure, it is either 9.2.0 or 17.10.1 i dont know where i can see version
* **OS**: windows 10
* **Scope (install, code, runtime, meta, other?)**: runtime, code
* **Module (and version) (if relevant)**: child_process.spawn

I want to use two processes (I'm spawning them using child process module). I want to use one of them to open input file and another one to open output file and then my application should read data from proc1, then process data, and finally write data to proc2 (which will save data as video file).

The problem is because the speed of proc1 reading data and speed of my application processing data and speed of proc2 saving data to file sometimes differ a lot. For example, proc1 reads chunk of data for 1ms, then my application processes it for 5ms and proc2 requires 18246ms to save data.

This is example how my code looks:

```JavaScript
proc1.stdout.on('data', data => {
  processData(data);
  proc2.stdin.write(data);
});
```

But, after about 1000 chunks of data, my PC becomes unresponsible and the system crashes, so I need to restart whole PC. I dont know much about async listeners, but something is telling me that I'm definitely doing this wrong. As I can tell, probably some internal or external buffer is filling much faster than emptying, so it leads to some sort of operating system memory leak, idk.

How can I properly attach event listeners and how can I prevent crashing? So, to clarify this, proc1 and proc2 and my app processes data with different speeds. How can I tell proc1 to wait until I process data and until I send it to proc2 and until proc2 processes data?

I suppose I need to attach callback listener to `proc2.stdin.write`, right? But I'm not sure what I need to do in that callback? How can I prevent proc1 from sending another chunk of data until proc2 is ready to receive another chunk?

Thanks you very much in advanced, sorry my english is broken.

**Comment by ORESoftware** [comment]
one first step might be

```js
proc1.stdout.on('data', data => {
  processData(data, function(err, d){
       proc2.stdin.write(d);
  });
});
```

and if you only want to write to proc2, when it's ready, you will need some signal from proc2.

the best thing to do might be to create some queue - perhaps on the filesystem. proc2 will read from that queue when it's ready. So proc2 will pull data, instead pushing data to proc2.

**Comment by wappevifi** [comment]
@ORESoftware I don't see how making `processData` asynchronous will solve the problem.

> If you only want to write to proc2, when it's ready, you will need some signal from proc2.

I think `proc2` is ready when it reads the very last byte of the chunk. The problem here is how to tell `proc1` not to send data until `proc2` reads the whole chunk. It is the main question here.

> The best thing to do might be to create some queue - perhaps on the filesystem.

Notice that `proc1` and my app take, in total, just few milliseconds to process data, but `proc2` takes few thousands milliseconds. My goal is to work continiously. It means, I want to be able to keep my app working for arbitrary long period of time. So, your idea about filesystem queue will not work, because the queue will grow constantly over time and, eventually, take up whole space on my hard disk.

Maybe it is worth mentioning that `proc1` will not send another chunk of data until I read the previous chunk (until my callback on `proc1.stdout.on('data')` is called). Also, `proc2` will not read the next chunk until it process the previous chunk. Therefore, I should somehow be able to deal with it, because it seems that I have everything I need, but I still don't know exactly which callbacks I need to attach in order to make it work as I want.

**Comment by ORESoftware** [comment]
@wappevifi well maybe someone who is more expert with Node.js streams can step in. But in general, if your disk can't handle the buffered data, not much else can, I would imagine.

**Comment by wappevifi** [comment]
> If your disk can't handle the buffered data, not much else can.

I think you misunderstand it. No data should be stored. Proc1 sends data, my app process it and send to proc2. Data is stored nowhere. So, data constantly flows and, theoretically, my app should be able to run forever. Adding data into queue really makes no sense, because it would require infinite amount of space, which is impossible. Data shouldn't be accumulated at all.

**Comment by ORESoftware** [comment]
Yeah the disk is not for storing the data. The disk is just for buffering the data. It sounds like the source is arriving faster than the destination can process it, so maybe you are blowing your Node.js memory.

**Comment by lpinca** [comment]
@wappevifi try to handle backpressure. Assuming that `processData()` works synchronously you can try with something like this:

```js
proc2.stdin.on('drain', () => proc1.stdout.resume());

proc1.stdout.on('data', data => {
  processData(data);
  if (proc2.stdin.write(data) === false) proc1.stdout.pause();
});
```

Another option is to extend `stream.Writable` and implement `_write`.

```js
const { Writable } = require('stream');

class MyWritable extends Writable {
  constructor(writer, options) {
    super(options);
    this.writer = writer;
  }

  _write(chunk, encoding, callback) {
    processData(chunk);
    this.writer.write(chunk, callback);
  }
}

const ws = new MyWritable(proc2.stdin);
proc1.stdout.pipe(ws);
```

The same applies upstream, you should stop writing to `proc1.stdout` when `false` is returned.

**Comment by wappevifi** [comment]
@lpinca Thanks a lot! I tried your code (the first one - using `pause` and `resume`) and my node is not crashing anymore.

I knew about `pause` and `resume`, but I thought they will cause some data chunks to be lost. Constructor of `proc.stdin` and `proc.stdout` is `net.Socket`, right? From the [docs](https://nodejs.org/api/net.html#net_socket_pause) it says

> Pauses the reading of data. That is, `data` events will not be emitted. Useful to throttle back an upload.

So, what does it exactly mean? Is it guaranteed that no data chunks will be lost, even if the socket is paused for a very long period of time? I tested my old program, and so far no data is lost, so is it right: will all data be processed, or some chunks may be lost?

Thank you very much for response.
